File: controllers/file-controller.js
import 'dotenv/config';
import fs from "fs";
import { FileModel } from '../models/file-model.js';
import { getRepository } from 'typeorm';
import path from 'path';
import { AppDataSource } from '../db/data-source.js';
import pkg from 'fluent-ffmpeg';
const { ffprobe } = pkg;
const fileController = {};

const getVideoDuration = (filePath) => {
  return new Promise((resolve, reject) => {
    ffprobe(filePath, (err, metadata) => {
      if (err) {
        return reject(err);
      }
      // The duration is in seconds
      const duration = metadata.format.duration;
      resolve(duration);
    });
  });
};


// * Function to upload files
fileController.UploadFile = async (req, res) => {
  if (!req.file) {
    return res.status(400).send('No file uploaded.');
  }

  try {
    const { filename, path: filePath } = req.file;
    const fileExtension = filename.split('.').pop();
    const fullFilePath = path.join(process.cwd(), filePath);

    console.log('File details:', fullFilePath, fileExtension);

    let totalLength = null;
    let Filetype;
    // Check if the file is a video by its extension
    const videoExtensions = ['mp4', 'mov', 'avi', 'mkv', 'webm'];
    const audioExtensions = ['mp3', 'wav', 'ogg', 'flac'];
    const imageExtensions = ['jpg', 'jpeg', 'png'];

    if (videoExtensions.includes(fileExtension.toLowerCase())) {
      // Use ffmpeg to get the video duration
      totalLength = await getVideoDuration(fullFilePath);
      console.log('Video duration:', totalLength);
      Filetype = 'video';
    }
    else if (audioExtensions.includes(fileExtension.toLowerCase())) {
      Filetype = 'audio';
    }
    else if (imageExtensions.includes(fileExtension.toLowerCase())) {
      Filetype = 'images';
    }
    else {
      return res.status(400).send('Invalid file type');
    }

    // Use the repository pattern to create and save the file details
    const fileRepository = AppDataSource.getRepository('FileModel'); // Ensure 'FileModel' matches your entity
    const fileDetails = fileRepository.create({
      filename,
      fileExtension,
      filePath: fullFilePath,
      fileType: Filetype,
      totalLength, // Store the total length of the video if available
    }); 

    await fileRepository.save(fileDetails); // Save to the database

    return res.send({
      message: 'File uploaded successfully!',
      fileName: filename,
      totalLength: totalLength || 'Not a video file',
    });
  } catch (error) {
    console.error('Error saving file details:', error);
    return res.status(500).send('An error occurred while uploading the file.');
  }
};

// * Function to get files
fileController.GetFiles = async (req, res) => {
  const FileType = req.query.type;

  if (!FileType) {
    return res.status(400).send("Missing file type query parameter");
  }

  if (FileType !== 'video' && FileType !== 'audio' && FileType !== 'images') {
    return res.status(400).send("Invalid file type query parameter");
  }

  const fileRepository = AppDataSource.getRepository(FileModel);


  if (FileType === 'video' || FileType === 'audio' || FileType === 'images') {
    const files = await fileRepository.find({ where: { fileType : FileType } });
    return res.send({"file_result" : files});
  }

  return res.status(500).send("An error occurred while fetching files");
};

// * Function to get the info of a video
fileController.GetVideoInfo = async (req, res) => {
  const fileId = req.query.id;
  if (!fileId) {
    return res.status(400).send("Missing video ID query parameter");
  }

  const fileRepository = AppDataSource.getRepository(FileModel);
  const videoFile = await fileRepository.findOne({ where: { id: fileId } });

  if (!videoFile) {
    return res.status(404).send("Video file not found");
  }

  return res.send({
    filename: videoFile.filename,
    totalLength: videoFile.totalLength || 'Not a video file',
    currentTrackAt: videoFile.currentTrackAt || 0,
  });
};

// * Function to stream video
fileController.StreamVideo = async (req, res) => {
  const range = req.headers.range;
  
  const fileId = req.query.id;
  console.log('Video ID:', fileId);
  if (!fileId) {
    return res.status(400).send("Missing video ID query parameter");
  }

  const fileRepository = AppDataSource.getRepository(FileModel);
  const videoFile = await fileRepository.findOne({ where: { id: fileId } });

  if (!videoFile) {
    return res.status(404).send("Video file not found");
  }

  const videoPath = videoFile.filePath;
  const currentTrackAt = videoFile.currentTrackAt || 0;


  const videoSize = fs.statSync(videoPath).size;
  const CHUNK_SIZE = 10 ** 6; // 1MB per chunk


  let start;
  if (range) {
    // If Range header exists, use it to calculate start
    start = Number(range.replace(/\D/g, ""));
  } else {
    start = Math.floor((currentTrackAt / 100) * videoSize);
  }
  const end = Math.min(start + CHUNK_SIZE - 1, videoSize - 1);

  // Create response headers
  const contentLength = end - start + 1;
  const headers = {
    "Content-Range": `bytes ${start}-${end}/${videoSize}`,
    "Accept-Ranges": "bytes",
    "Content-Length": contentLength,
    "Content-Type": "video/mp4",
  };

  res.writeHead(206, headers);

  const videoStream = fs.createReadStream(videoPath, { start, end });
  videoStream.pipe(res);

};

// * Function to save the current time of a video
fileController.SaveCurrentTime = async (req, res) => {
  const fileId = req.body.videoId;
  const currentTime = req.body.currentTime;

  if (!fileId || !currentTime) {
    return res.status(400).send("Missing video ID or time query parameter");
  }

  await saveCurrentTime(fileId, currentTime);
  return res.send("Current time saved successfully");
};

const saveCurrentTime = async (fileId, currentTime) => {
  const fileRepository = AppDataSource.getRepository(FileModel);
  const videoFile = await fileRepository.findOne({ where: { id: fileId } });

  if (videoFile) {
    videoFile.currentTrackAt = currentTime; 
    await fileRepository.save(videoFile);
  }
};



// ? Export the fileController
export { fileController };

File: backend/src/app/v1/FileOperations/api/fileController.py
from fastapi import UploadFile, File, HTTPException, Depends, Request
from fastapi.responses import StreamingResponse
from sqlmodel import Session, select
import os
import aiofiles
import ffmpeg
from ..models.models import FileModel
from src.database.db import get_session

async def get_video_duration(file_path: str) -> float:
    probe = ffmpeg.probe(file_path)
    return float(probe['format']['duration'])


async def uploadFile(file: UploadFile = File(...), title: str = None, db: Session = Depends(get_session)):
    if not title:
        raise HTTPException(status_code=400, detail="Title is required")
    
    file_ext = file.filename.split(".")[-1]
    file_path = f"storage/{file.filename}"
    os.makedirs("storage", exist_ok=True)
    
    async with aiofiles.open(file_path, "wb") as buffer:
        while content := await file.read(1024):
            await buffer.write(content)
    
    file_type = "unknown"
    if file_ext in ["mp4", "mov", "avi", "mkv", "webm"]:
        file_type = "video"
        total_length = await get_video_duration(file_path)
    elif file_ext in ["mp3", "wav", "ogg", "flac"]:
        file_type = "audio"
        total_length = None
    elif file_ext in ["jpg", "jpeg", "png"]:
        file_type = "image"
        total_length = None
    else:
        raise HTTPException(status_code=400, detail="Invalid file type")
    
    new_file = FileModel(title=title, filename=file.filename, file_extension=file_ext, file_path=file_path, file_type=file_type, total_length=total_length)
    db.add(new_file)
    db.commit()
    return {"message": "File uploaded successfully!", "file_name": file.filename, "total_length": total_length or "Not a video file"}


def getFiles(file_type: str, db: Session = Depends(get_session)):
    if file_type not in ["video", "audio", "image"]:
        raise HTTPException(status_code=400, detail="Invalid file type")
    files = db.exec(select(FileModel).where(FileModel.file_type == file_type)).all()
    return {"file_result": files}


def getVideoInfo(file_id: int, db: Session = Depends(get_session)):
    video_file = db.get(FileModel, file_id)
    if not video_file:
        raise HTTPException(status_code=404, detail="Video file not found")
    return {"filename": video_file.filename, "total_length": video_file.total_length or "Not a video file", "currentTrackAt": video_file.current_track_at or 0}


def streamVideo(file_id: int, request: Request, db: Session = Depends(get_session)):
    video_file = db.get(FileModel, file_id)
    if not video_file:
        raise HTTPException(status_code=404, detail="Video file not found")
    
    file_path = video_file.file_path
    file_size = os.path.getsize(file_path)
    chunk_size = 1024 * 1024  # 1MB chunks

    range_header = request.headers.get("range")
    start = 0
    end = file_size - 1

    if range_header:
        range_values = range_header.replace("bytes=", "").split("-")
        start = int(range_values[0]) if range_values[0] else 0
        end = int(range_values[1]) if len(range_values) > 1 and range_values[1] else end

    end = min(end, file_size - 1)  # Ensure end is within bounds

    def file_iterator(start_pos, end_pos):
        with open(file_path, "rb") as f:
            f.seek(start_pos)
            while start_pos <= end_pos:
                data = f.read(min(chunk_size, end_pos - start_pos + 1))
                if not data:
                    break
                yield data
                start_pos += len(data)

    return StreamingResponse(
        file_iterator(start, end),
        status_code=206,
        headers={
            "Content-Range": f"bytes {start}-{end}/{file_size}",
            "Accept-Ranges": "bytes",
            "Content-Length": str(end - start + 1),
            "Content-Type": "video/mp4",
        },
    )


def saveVideoTime(file_id: int, current_time: float, db: Session = Depends(get_session)):
    video_file = db.get(FileModel, file_id)
    if not video_file:
        raise HTTPException(status_code=404, detail="Video file not found")
    video_file.current_track_at = current_time
    db.add(video_file)
    db.commit()
    return {"message": "Current time saved successfully"}


def showImage(file_id: int, db: Session = Depends(get_session)):
    image_file = db.get(FileModel, file_id)
    if not image_file:
        raise HTTPException(status_code=404, detail="Image file not found")
    return StreamingResponse(open(image_file.file_path, "rb"), media_type="image/jpeg")

def streamAudio(file_id: int, db: Session = Depends(get_session)):
    audio_file = db.get(FileModel, file_id)
    if not audio_file:
        raise HTTPException(status_code=404, detail="Audio file not found")
    return StreamingResponse(open(audio_file.file_path, "rb"), media_type="audio/mpeg")
